### PR TITLE
Feature/cafe list

### PR DESCRIPTION
**Pull request by Sprokopyak**
Add cafe list

### CODE REVIEW

File: src/app/cafe/add-cafe/add-cafe.component.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  cafeTypes = CAFE_TYPES;
+  latitude: number;
+  longitude: number;
+  tables = [];
+  currentUpload: Upload;
+  mainImgSrc: object;
+  gallery = [];
+  progress$: Observable<number>;
+
+  @ViewChild('search') public searchElementRef: ElementRef;
```
Comment (GAndriyS) [inline src/app/cafe/add-cafe/add-cafe.component.ts]: no need to use public keyword

File: src/app/cafe/add-cafe/add-cafe.component.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+  constructor(
+    private _mapsAPILoader: MapsAPILoader, 
+    private _ngZone: NgZone, 
+    private _fb: FormBuilder,
+    public imageUploadService: ImageUploadService,
```
Comment (GAndriyS) [inline src/app/cafe/add-cafe/add-cafe.component.ts:37]: no need to use public keyword

File: src/app/cafe/add-cafe/add-cafe.component.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  onAddTables(tablesNumber, visitorsNumber) {   
+    if (tablesNumber.value !== '' && visitorsNumber.value !== '') {
+
+      this.tables.push({ tablesNumber: parseInt(tablesNumber.value), visitorsNumber: parseInt(visitorsNumber.value), booked: 0 });
+      tablesNumber.value = '';
+      visitorsNumber.value = '';
+    }
+  }
+
+  removeGalleryImg(img): void {
```
Comment (GAndriyS) [inline src/app/cafe/add-cafe/add-cafe.component.ts]: decide to specify or don't :void, be consistent across application

File: src/app/cafe/add-cafe/add-cafe.component.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    this.imageUploadService.removeImg(img.fullPath, img.thumbnailPath);
+    let index = this.gallery.indexOf(img);
+    this.gallery = this.gallery.filter((el, i) => i !== index);
+  }
+
+  removeTables(table: any): void {
+    let index = this.tables.indexOf(table);
```
Comment (GAndriyS) [inline src/app/cafe/add-cafe/add-cafe.component.ts]: use const as often as possible

File: src/app/core/cafe-service/cafe.service.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import { Observable } from 'rxjs';
+import { Star } from '../models/rating.model';
+import { Cafe } from '../models/cafe.model';
+
+import { MatDialog } from '@angular/material';
+import { MessageDialog } from '../../commons/message-dialog/message-dialog.component';
+
+@Injectable()
+export class CafeService {
+  cafeRef: AngularFirestoreCollection<Cafe> = this._afs.collection("cafes");
```
Comment (GAndriyS) [inline src/app/core/cafe-service/cafe.service.ts]: inconsistent double quotes

File: src/app/core/core.module.ts [hunk 6]
```diff
@@ -1,21 +1,33 @@
...
+import { AngularFirestoreModule } from "angularfire2/firestore";
+import { AngularFireAuthModule } from "angularfire2/auth";
+import { AngularFireDatabaseModule } from "angularfire2/database";
+import { AngularFireDatabase, FirebaseListObservable } from "angularfire2/database-deprecated";
 
-import { AuthService } from './auth-service/auth.service';
-import { AdminGuard } from './admin-guard/admin.guard';
-import { LoginGuard } from './login-guard/login.guard';
+import { AuthService } from "./auth-service/auth.service";
+import { CafeService } from "./cafe-service/cafe.service";
```
Comment (GAndriyS) [inline src/app/core/core.module.ts:9]: again inconsistent ' / "

File: src/app/core/image-upload/image-upload.service.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+export class ImageUploadService {
+  file;
+  private _mainPhotos: string = '/mainPhotos';
+  private _galaryPhotos: string = '/galaryPhotos';
+  completed$ = new Subject<Upload>();
+  uploading$ = new Subject<number>();
+  completedMulti$ = new Subject<Upload>();
+  selectedFiles: FileList;
+  uniqueId = Math.random().toString(36).substring(2) + (new Date()).getTime().toString(36);
+  thumbnailStorageUrl = 'https://firebasestorage.googleapis.com/v0/b/easy-book-2fcf6.appspot.com/o/';
```
Comment (GAndriyS) [inline src/app/core/image-upload/image-upload.service.ts:22]: group private and public props

File: src/app/core/infinity-scroll/infinity-scroll.service.ts [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+@Injectable()
+export class InfinityScrollService {
+
+  // Source data
+  private _done = new BehaviorSubject(false);
+  private _loading = new BehaviorSubject(false);
+  private _data = new BehaviorSubject([]);
+
+  private query: QueryConfig;
```
Comment (GAndriyS) [inline src/app/core/infinity-scroll/infinity-scroll.service.ts]: missed underscore

File: src/app/core/infinity-scroll/infinity-scroll.service.ts [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+      return ref
+              .orderBy(this.query.field, this.query.reverse ? 'desc' : 'asc')
+              .limit(this.query.limit)
+    })
+
+    this.mapAndUpdate(first)
+
+    // Create the observable array for consumption in components
+    this.data = this._data.asObservable()
+    .pipe(
```
Comment (GAndriyS) [inline src/app/core/infinity-scroll/infinity-scroll.service.ts]: optimize code formatting here

File: src/app/core/infinity-scroll/infinity-scroll.service.ts [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    if (current.length) {
+      return this.query.prepend ? current[0].doc : current[current.length - 1].doc 
+    }
+    return null
+  }
+
+
+  // Maps the snapshot to usable format the updates source
+  private mapAndUpdate(col: AngularFirestoreCollection<any>) {
+console.log(col);
```
Comment (GAndriyS) [inline src/app/core/infinity-scroll/infinity-scroll.service.ts]: missed console.log

File: src/app/core/infinity-scroll/infinity-scroll.service.ts [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+    if (this._done.value || this._loading.value) { return };
+
+    // loading
+    this._loading.next(true)
+
+    // Map snapshot with doc ref (needed for cursor)
+    return col.snapshotChanges()
+.pipe(
+      tap(arr => {
```
Comment (GAndriyS) [inline src/app/core/infinity-scroll/infinity-scroll.service.ts]: code formatting

File: src/app/core/infinity-scroll/infinity-scroll.service.ts [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        console.log(arr);
```
Comment (GAndriyS) [inline src/app/core/infinity-scroll/infinity-scroll.service.ts]: missed console.log

File: src/app/core/infinity-scroll/infinity-scroll.service.ts [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+        // no more values, mark done
+        if (!values.length) {
+          this._done.next(true)
+        }
+    }),
+    take(1)
+  )
+  
+  .subscribe(val=>{
+    console.log(val);
```
Comment (GAndriyS) [inline src/app/core/infinity-scroll/infinity-scroll.service.ts]: ????

File: src/app/directives/same-height.directive.ts [hunk 14]
```diff
@@ -0,0 +1,48 @@
+import {
+  Directive,
+  ElementRef,
+  AfterViewChecked,
+  Input,
+  HostListener
+} from "@angular/core";
+
+@Directive({
+  selector: "[samehHeight]"
```
Comment (GAndriyS) [inline src/app/directives/same-height.directive.ts]: " again

File: src/app/directives/scrollable.directive.ts [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+      if (top > height - offset - 1) {
+        this.scrollPosition.emit('bottom')
+      }
+
+      // emit top event
+      if (top === 0) {
+        this.scrollPosition.emit('top')
+      }
+
+    } catch (err) {}
```
Comment (GAndriyS) [inline src/app/directives/scrollable.directive.ts]: no need in try catch if there are no error handling

File: src/app/home/home.component.ts [hunk 16]
```diff
@@ -1,15 +1,100 @@
...
+  startAt = new Subject();
+  endAt = new Subject();
+  cafeTypes = CAFE_TYPES;
+  cafeTypeFilter$ = new BehaviorSubject(null);
+  ratingFilter$ = new BehaviorSubject(null);
+  freeTablesFilter$ = new BehaviorSubject(null);
+
+  constructor(
+    private _cafeService: CafeService,
+    private afs: AngularFirestore,
```
Comment (GAndriyS) [inline src/app/home/home.component.ts]: underscore for private props

File: src/app/home/home.component.ts [hunk 17]
```diff
@@ -1,15 +1,100 @@
...
+    public page: InfinityScrollService
```
Comment (GAndriyS) [inline src/app/home/home.component.ts]: no need in public keyword

**Review by GAndriyS — CHANGES_REQUESTED** [review]

File: src/app/home/home.component.ts (hunk 17 above)
Comment (Sprokopyak) [reply src/app/home/home.component.ts]: I use this service in html

File: src/app/cafe/add-cafe/add-cafe.component.ts (hunk 2 above)
Comment (Sprokopyak) [reply src/app/cafe/add-cafe/add-cafe.component.ts:37]: I use this service in html

**Pull request merged**